[PM-6826] Vault filter refresh

nick-livefront opened the pull request:
## 🎟️ Tracking

<!-- Paste the link to the Jira or GitHub issue or otherwise describe / point to where this change is coming from. -->
[PM-6826](https://bitwarden.atlassian.net/browse/PM-6826)

## 📔 Objective

<!-- Describe what the purpose of this PR is, for example what bug you're fixing or new feature you're adding. -->
- Add the ability to filter ciphers within the Extension Vault
  - Behind `ExtensionRefresh` feature flag
- Available filters: `Vault`, `Collection`, `Folder`, & `Type`

### Open Thoughts

- I'm pondering if there are some efficiency gains to be had but I wanted to get this PR up. I'll see if there are some other approaches I could take with the observables here. Though I'm open to ideas!
- I needed to alter the `ChipSelectComponent` in order for the integration to work here. I'll leave individual comments as to _why_.
- ~~This will have conflicts with @Jingo88's https://github.com/bitwarden/clients/pull/9343, I'll keep an eye out on changes there for any conflicts/implementation changes!~~ Jason's branch was merged and I've fixed conflicts!

## 📸 Screenshots

<!-- Required for any UI changes; delete if not applicable. Use fixed width images for better display. -->


|General Functionality|Deactivated Org|
|---------------------|----------------|
|<img src="https://github.com/bitwarden/clients/assets/125900171/a512b6ce-e694-41fe-8ec6-1ad6f04145bd" width="376" />|<img width="376" alt="Screenshot 2024-05-24 at 3 41 28 PM" src="https://github.com/bitwarden/clients/assets/125900171/c6281203-fc73-4232-8e1b-ccef2acbd4ee">|

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes


[PM-6826]: https://bitwarden.atlassian.net/browse/PM-6826?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nick-livefront commented on libs/components/src/chip-select/chip-select.component.ts:
```diff
@@ -108,7 +108,7 @@ export class ChipSelectComponent<T = unknown> implements ControlValueAccessor {
   private findOption(tree: ChipSelectOption<T>, value: T): ChipSelectOption<T> | null {
     let result = null;
-    if (tree.value === value) {
+    if (tree.value !== null && tree.value === value) {
```
Reason for change: 

When my initial value was null and that the `tree` value was also null, it would return the `tree` resulting in `this.selectedOption` being set to a truthy value. 
So the empty state is shown as selected:
<img width="383" alt="Screenshot 2024-05-24 at 5 26 07 PM" src="https://github.com/bitwarden/clients/assets/125900171/eece81fa-d34a-4fc3-b1bc-75e6b9a3b1a2">

nick-livefront commented on libs/components/src/chip-select/chip-select.component.ts:
```diff
@@ -183,7 +183,7 @@ export class ChipSelectComponent<T = unknown> implements ControlValueAccessor {
     }
 
-    this.notifyOnChange(option?.value);
+    this.notifyOnChange(option?.value ?? null);
```
The optional chaining here is calling `notifyOnChange` with `undefined` rather than `null`, which resulted in logic issues when `null` is used for empty states.

The inverse could be argued, when `undefined` is used for an empty state. In my opinion `null` is a more natural use case for these scenarios but I'm open to the discussion!

nick-livefront commented:
I'm moving this to a draft, I realized that if selecting one filter negates another one the filter is still being applied. I have to rework how I'm handling the filters. 

I'm still open to feedback in the meantime!

shane-melton commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.ts:
```diff
@@ -0,0 +1,312 @@
+
+  /** Observable of the available cipher types */
+  cipherTypes$: Observable<ChipSelectOption<string | CipherType>[]> = combineLatest([
+    this.vaultSettingsService.showCardsCurrentTab$,
```
issue: These `show*CurrentTab$` settings are meant to only apply to the autofill list section and determine if Cards/Identities should show up there in addition to Logins.

Otherwise, I believe we should always show all cipher types in the filter chip regardless of the settings.

nick-livefront replied:
Removed! [7d0e096](https://github.com/bitwarden/clients/pull/9365/commits/7d0e0967f53553ec3936060ac02909f1c169560a)

shane-melton commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.ts:
```diff
@@ -0,0 +1,312 @@
+              if (!org.enabled) {
+                // Show a warning icon if the organization is deactivated
+                icon = "bwi-exclamation-triangle tw-text-danger";
+              } else if (org.planProductType === 1) {
```
nit (non-blocking): Can we use the `ProductType` enum here instead of `1`?

nick-livefront replied:
[53e4e4b](https://github.com/bitwarden/clients/pull/9365/commits/53e4e4baa468a4cfe9e61bbd5be6d5aaebdb2f77)

shane-melton commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.ts:
```diff
@@ -0,0 +1,312 @@
+    this.organizationService.memberOrganizations$.pipe(
+      map((orgs) => orgs.sort(Utils.getSortFunction(this.i18nService, "name"))),
+      map((orgs) => {
+        if (orgs.length) {
```
nit (non-blocking): Smallest of nits/suggestion, but we could invert this `if` to reduce the nesting here a bit.

shane-melton commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.ts:
```diff
@@ -0,0 +1,312 @@
+  };
+
+  /** Observable of the available cipher types */
+  cipherTypes$: Observable<ChipSelectOption<string | CipherType>[]> = combineLatest([
```
question: Why do we have `string | CipherType[]` here for the `ChipSelectOption`?

nick-livefront replied:
I'm not sure why I had to add that but I remember jumping hoops for typescript. But this was eliminated by making the list of ciphers constant

shane-melton commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.ts:
```diff
@@ -0,0 +1,312 @@
+      const organizationId = filters.organizationId;
+
+      // Get all stored collections
+      const allCollections = await this.collectionService.getAllDecrypted();
```
suggestion: We can use the `collectionService.decryptedCollections$` observable instead of `getAllDecrypted()`

shane-melton commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.spec.ts:
kudos: As always, thanks for the tests!

shane-melton commented on apps/browser/src/vault/popup/services/vault-popup-items.service.ts:
```diff
@@ -69,14 +73,24 @@ export class VaultPopupItemsService {
 
+  private _filteredCipherList$: Observable<CipherView[]> = combineLatest([
+    this._cipherList$,
+    this.vaultPopupListFiltersService.filters$,
```
thought: Random idea, not necessarily better, but what if we had a `filterFunction$` observable on the filter service and subscribe to that. 

I think it might help avoid some of the awkwardness around subscribing to a service value and then passing it back into said service. What do you think?

nick-livefront replied:
This might be one of my favorite changes because I don't like it and like it. It feels kind of odd to me that the service would have an observable function but the actual implementation is exactly what we need so I'm all for it! 

The overall structure is an improvement, I like it. Let's roll. [acf4d97](https://github.com/bitwarden/clients/pull/9365/commits/acf4d9791776269fd5c6603d51a612fc3e1cdaf7)

shane-melton replied:
Awesome, thanks for trying it out!

shane-melton reviewed (changes requested):
This is looking great! I know its still in Draft but I added a few comments on what you have so far!

nick-livefront commented on apps/browser/src/vault/popup/components/vault-v2/vault-list-filters/vault-list-filters.component.ts:
```diff
@@ -0,0 +1,50 @@
+  ) {}
+
+  ngOnInit(): void {
+    this.filterForm.valueChanges.pipe(takeUntil(this.destroy$)).subscribe((formValue) => {
```
Noted for next time, I nixed this change based on moving the filterForm into the service 🚀

nick-livefront commented:
@shane-melton I was running into an error when I added `FormBuilder` to the filter service tests. The internet lead me to adding `"jest-preset-angular/setup-jest"` to the `test.setup.ts` file. This caused some other, unrelated tests to break. I fixed them here: [820d6ab](https://github.com/bitwarden/clients/pull/9365/commits/820d6ab0e0d3710bbeefa4e7070f8c04b7feeb87) but I also pulled them into a separate PR #9412 just so they're called out separately.

shane-melton commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.ts:
```diff
@@ -0,0 +1,364 @@
+      folders.sort(Utils.getSortFunction(this.i18nService, "name"));
+      let arrangedFolders = folders;
+
+      const noFolder = folders.find((f) => f.id === null);
```
♻️ This no folder == null Id thing seems kind of annoying to work around, especially since its already added by the service. It's forcing us to iterate the `folders` array 2 more times here when we should just be able to append it to the end. 

The team should probably look into cleaning this up in the future; but for now this should be fine.

shane-melton commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.ts:
```diff
@@ -0,0 +1,364 @@
+  /**
+   * UI form for all filters
+   */
+  filterForm = this.formBuilder.group<PopupListFilter>(INITIAL_FILTERS);
```
💭 I'm kind of torn on adding the actual `FormGroup` to the service. On the one hand, it does avoid the need for a separate `updateFilters` method on the service and having to manually keep another `FormGroup` in sync; so I definitely see its value. But, it also feels unfortunate that we now have this dependency on Angular in this service AND we have to subscribe to `valueChanges` in the constructor.

shane-melton replied:
❓ Do we need to be concerned with reseting this `formGroup` since its living in a service that might outlive the new Vault tab and have stale values?

nick-livefront replied:
I can see both sides, what pushed me to move it to the service was creating a single data model that would drive the logic & the UI. When I had the `filterForm` in the component, I was writing more logic to keep the `filterForm` in sync with the `filter$` observable. 

> Do we need to be concerned with reseting this formGroup since its living in a service that might outlive the new Vault tab and have stale values?

I think that's an easy add, I added a reset to the form when the filter component is destroyed. [505e811](https://github.com/bitwarden/clients/pull/9365/commits/505e811e007f535f7f6ef53d2bb2bdbadcbcf4c5)

shane-melton commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.ts:
```diff
@@ -0,0 +1,364 @@
+  ) {
+    this.filterForm
+      .get("organization")
+      .valueChanges.subscribe(this.validateOrganizationChange.bind(this));
```
❓ I'm not entirely certain on the lifetime of this service, but maybe we should have a `takeUntilDestroyed()` operator here?

nick-livefront replied:
👍 [51db8dc](https://github.com/bitwarden/clients/pull/9365/commits/51db8dc114c33b96e68cacc219ac32377936ae5f)

shane-melton commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.ts:
```diff
@@ -0,0 +1,364 @@
+  /**
+   * Observable whose value is a function that filters an array of `CipherView` objects based on the current filters
+   */
+  filterFunction$: Observable<(ciphers: CipherView[]) => CipherView[]> = this.filters$.pipe(
```
👍 I really like how this turned out actually! Fully embracing the [first class citizenship](https://en.wikipedia.org/wiki/First-class_function) of functions in JS haha

shane-melton reviewed (commented):
Looking great! Just a few more thoughts/questions.

nick-livefront commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.ts:
```diff
@@ -0,0 +1,364 @@
+      const orgCiphers = cipherViews.filter((c) => c.organizationId === organizationId);
+
+      // Return only the folders that have ciphers within the filtered organization
+      return folders.filter((f) => orgCiphers.some((oc) => oc.folderId === f.id) || f.id === null);
```
I don't believe so! Removed [c3af6b1](https://github.com/bitwarden/clients/pull/9365/commits/c3af6b18179bcb5e1ad248fb91f1fca30993138b)

shane-melton reviewed (commented):
One last small question, but I think we're pretty much there!

nick-livefront commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.ts:
```diff
@@ -87,10 +88,12 @@ export class VaultPopupListFiltersService {
+    destroyRef: DestroyRef,
   ) {
     this.filterForm
       .get("organization")
```
👍 [f04bf5c](https://github.com/bitwarden/clients/pull/9365/commits/f04bf5ccba668db22cba88d37a1dfc300e83519e)

nick-livefront commented on apps/browser/src/vault/popup/services/vault-popup-list-filters.service.ts:
```diff
@@ -87,10 +88,12 @@ export class VaultPopupListFiltersService {
     private i18nService: I18nService,
     private collectionService: CollectionService,
     private formBuilder: FormBuilder,
+    destroyRef: DestroyRef,
```
Good call out, yes it does use `inject` under the hood. I was struggling to get the unit tests to run without passing the `destroyRef` manually.

Turns out that is just because the test wasn't running within the `TestBed`. Using the `new VaultPopupListFilterService(..)` syntax was the culprit. Switched `TestBed.inject(VaultPopupListFilterService)` and all is well.

[4d02197](https://github.com/bitwarden/clients/pull/9365/commits/4d0219706bb2eddcbd54a03c617512e55450acff)

nick-livefront commented:
@willmartian When you have the time could you give this a review? I made some changes here to the `ChipSelect` component.

willmartian reviewed (approved):
CL changes LGTM, thanks @nick-livefront!
